main.go

```diff
@@ -200,6 +198,59 @@ func getTrailheadID(w http.ResponseWriter, userAlias string) string {
 	return userAlias
 }
 
+func doTrailheadAuraCallout(apexAction string, pageURI string) trailhead.Data {
+	// If config has been retrieved, try aura call
+	if 0 != len(auraConfig.FwUid) {
```

Comment:
I was talking with a friend of mine (who knows way more about Go than I do) about handling this and while updating the config for this instance using a global var works locally and probably on the Heroku server, it's not preferred for a server application. I think the problem that could arise is serving multiple instances of the application at once. 

AFAIK, Heroku serves one instance of the application once the dyno is spun up, since the dyno goes to sleep whenever it hasn't been pinged in a while. The way you have this built currently, when the application is first spun up it'll need to get the `fwuid` value from your `updateAuraProfileAppConfig()` method and cache it. All subsequent requests should then reference that value. So in theory, and in my local testing, it works perfectly. 

IMO, this solution should work fine enough for one instance of the application run on Heroku. And even if there were multiple requests, or Go routines, the way this is built would just get the value the first time, so I say we leave this as is for now and see how it works. The only thing I would say we'd think about is on line [208](https://github.com/meruff/go-trailhead-leaderboard-api/pull/9/files#diff-7ddfb3e035b42cd70649cc33393fe32cR208) where it checks if `Actions` is not `nil`. What if `Actions` IS `nil`? We should add an else that would do a callout to `updateAuraProfileAppConfig` to update the `fwuid` value. In this case, the server was up, with a valid `fwuid`, but then that value changed in the middle of the instance so it became invalid, thus a new value is needed again. There's probably a better way to implement that instead of an else with duplicate calls to code, but I'll let you decide if you wanna refactor `doTrailheadAuraCallout` now or just leave it. I can look at it later and see if we can simplify that.

Reply:
I agree with you in that a multi-instance container each would have to store this value in memory but it's preferable to a database value since each instance would need the refresh. You'd need a cache layer or a notification system to synchronize the value. The performance hit of this call is low with an immediate recovery. 

With regards to the flow, if the ID is present and the call fails, it won't return and will update the ID and then retry the call. I thought about using recursion instead of duplicating the 3 lines but the duplication prevents an infinite loop bug from ever existing. Would only want to retry once in case there was some other problem. The code will recover as-is if the FwUID changes without the instance having to be restarted. I didn't explicitly use the else because this flow will get the config "just in time" on the first call. So if it exists and fails it will just continue like it was the first call to the instance.